### DISCUSSION TITLE
How can I make authenticated requests to NextJS API from a different FE client? (Shared auth with chrome extension)

### DISCUSSION THREAD

**Discussion by jellohouse** [supabase/supabase#23297] [Questions] [answered]
Hi,

I have setup a fully functional NextJS web app with Supabase SSR authentication. Everything works great!

But now I need to make authenticated requests to my NextJS API protected routes from a different static FE client (in my case a chrome extension)

I'm able to login the user to my extension using simply supabase js.

But i'm having issues trying to make requests to my NextJS API route handlers that require user authentication.

I know Supabase SSR uses cookies on the web app. Is there a way I should create a cookie in my chome extension?

I know I have access to an access_token in my chrome extension... any way I can use that?

Just wondering what is the best approach here.
Thanks in advance!

**Comment by jellohouse** [comment]
Ok so little update - I tried using the existing authenticated session from the web app (cookies) in the browser extension. So they share the same session.

The requests to the NextJS API are good. But for some reason after a few hours I get logged out of both.

Not too sure what the problem might be...

Here is how i initiate the Supabase client in my extension backgrounds script. 

```javascript
const storageAdapterBG = {
  getItem: async (key) => {
    const cookie = await chrome.cookies.get({ url: process.env.PLASMO_PUBLIC_WEB_APP_URL, name: key });
    return decodeURIComponent(cookie?.value);
  },

  setItem: async (key, value) => {
    await chrome.cookies.set({
      url: process.env.PLASMO_PUBLIC_WEB_APP_URL,
      name: key,
      value: value,
    });
  },

  removeItem: async (key) => {
    await chrome.cookies.remove({ url: process.env.PLASMO_PUBLIC_WEB_APP_URL, name: key });
  },
};

export const supabaseBG = createClient(
  process.env.PLASMO_PUBLIC_SUPABASE_URL,
  process.env.PLASMO_PUBLIC_SUPABASE_KEY,
  {
    auth: {
      storage: storageAdapterBG,
      autoRefreshToken: true,
      detectSessionInUrl: true,
      persistSession: true,
    },
  }
);
```

What could be causing the issue of getting logged out after a while?

Both the web app and extension are using the same cookies...

**Reply by GaryAustin1** [reply]
Don't know your environments at all.  But if "after a while" corresponds with the jwt expiration time frame (default 1 hour) then you don't have a supabase-js client refreshing, or the client that is refreshing it not updating the cookie or the user's of the cookie are not knowing to load the new updated cookie.

**Reply by jellohouse** [reply]
The issue was I forgot to do `encodeURIComponent(value)` on the setItem function

**Comment by jellohouse** [comment, marked as answer]
Actually I found the issue - I forgot to do encodeURIComponent on the setItem function. 

And I can't believe there was also a solution online I found after I came up with this myself. Here is what i found: https://github.com/csenio/supabase-auth-helper-webextension

Here is the corrected code for anyone interested:

```javascript
const storageAdapterBG = {
  getItem: async (key) => {
    const cookie = await chrome.cookies.get({ url: process.env.PLASMO_PUBLIC_WEB_APP_URL, name: key });
    return decodeURIComponent(cookie?.value);
  },

  setItem: async (key, value) => {
    await chrome.cookies.set({
      url: process.env.PLASMO_PUBLIC_WEB_APP_URL,
      name: key,
      value: encodeURIComponent(value),
    });
  },

  removeItem: async (key) => {
    await chrome.cookies.remove({ url: process.env.PLASMO_PUBLIC_WEB_APP_URL, name: key });
  },
};

export const supabaseBG = createClient(
  process.env.PLASMO_PUBLIC_SUPABASE_URL,
  process.env.PLASMO_PUBLIC_SUPABASE_KEY,
  {
    auth: {
      storage: storageAdapterBG,
      autoRefreshToken: true,
      detectSessionInUrl: true,
      persistSession: true,
    },
  }
);
```

**Reply by jamesleeht** [reply]
Thanks for the answer - this is super helpful. There was another discussion related to authenticating on the extension itself but that was not what I wanted (I wanted to share the session with the webapp).

How did you arrive at this answer? Especially the config part:

```
export const supabaseBG = createClient(
  process.env.PLASMO_PUBLIC_SUPABASE_URL,
  process.env.PLASMO_PUBLIC_SUPABASE_KEY,
  {
    auth: {
      storage: storageAdapterBG,
      autoRefreshToken: true,
      detectSessionInUrl: true,
      persistSession: true,
    },
  }
);
```

I didn't really see this mentioned in the docs.

**Reply by jamesleeht** [reply]
There seem to be a few issues with this implementation:

1. Authenticating using socials will result in cookies that have a different key. (suffixed with .0 or .1).
2. Remove item seems to be called randomly causing the token to be removed from cookies.
